### PR TITLE
[Issue #151] feat: using real prices in the dashboard

### PR DESCRIPTION
**Pull request by chedieck**
Description:
Now we  use the price at the time of each XEC and BCH tx to calculate the dashboard data. Solves #151 

Test plan:
1. Log in as dev2@paybutton.org,
2. Create a paybutton for some XEC/BCH address that you can more easily check the last few payments value using the price at the moment,
3. Log out and back in (as dev2@paybutton.org) to sync the transactions for that address,
4. The dashboard information should reflect the expected values.

Remarks:
The following is also being fixed:
- The pipeline for generating the dashboard payment data was weirdly modifying the transaction objects so that `transaction.amount` holds the value instead of the amount transactioned.
- Upserting many price rows asynchronously was causing a race condition in the database.

Depends on:
- [x] #295

### CODE REVIEW

**Comment by chedieck** [comment]
There are now also some changes related to conciliating merge conflicts with #296

**Review by ScottMcDermid — CHANGES_REQUESTED** [review]
I receive an error when I first log in. Or if I delete transactions, then log in.

```
error - (api)/services/transactionService.ts (51:45) @ Module.getTransactionValue
Error: txid215, ts1667692032 Missing price for transaction.
  49 |     cad: new Prisma.Decimal(0)
  50 |   }
> 51 |   if (transaction.prices.length !== 2) throw new Error(`txid${transaction.id}, ts${transaction.timestamp} ${RESPONSE_MESSAGES.MISSING_PRICE_FOR_TRANSACTION_400.message}`)
     |                                             ^
  52 |   for (const p of transaction.prices) {
  53 |     if (p.price.quoteId === USD_QUOTE_ID) {
  54 |       ret.usd = ret.usd.plus(p.price.value.times(transaction.amount))
```

Tested a payment, and it was detected as expected. Just one code comment.

**Review by Klakurka — CHANGES_REQUESTED** [review]
I get the same error as Scott on initial sync:

> paybutton-dev    | error - (api)/services/transactionService.ts (51:45) @ Module.getTransactionValue

but after waiting a bit everything started working.

While the error is happening, the page won't load which is probably an issue.

File: services/transactionService.ts [hunk 1]
```diff
@@ -42,6 +43,23 @@ export async function getTransactionAmount (transaction: BCHTransaction.AsObject
   )
 }
 
+export async function getTransactionValue (transaction: TransactionWithAddressAndPrices): Promise<QuoteValues> {
+  const ret: QuoteValues = {
+    usd: new Prisma.Decimal(0),
+    cad: new Prisma.Decimal(0)
+  }
+  if (transaction.prices.length !== 2) throw new Error(`txid${transaction.id}, ts${transaction.timestamp} ${RESPONSE_MESSAGES.MISSING_PRICE_FOR_TRANSACTION_400.message}`)
```
[the remark this answers is not in the document]
Comment (chedieck) [reply services/transactionService.ts]: It is 2 because we expect the price both in USD and in CAD, and since we always expect to have both of them, checking for equality against `2` is ideal, because no more or no less would be right.

In any case, writing the hardcoded `2` is indeed a bad practice, so I created a constant `N_OF_QUOTES` for it (check the second to last commit)

**Comment by chedieck** [comment]
About the error:

_TL;DR: Fixed in the last commit. The error shouldn't appear anymore & the page shouldn't appear blank, taking long to load._


I wrote the code in such a way that it should first sends all the requests for the price information, and only after all the responses are received this information is uploaded in the DB, all at once. This avoid condition race when the data is being written to the DB and also speeds up the process, since we do not need to wait each response to send the next request.

However, it was actually doing the opposite of that: sending one request at a time, and therefore taking really long to finish the sync. Besides that, it would try to fetch the prices before the sync was finished, so the prices would be missing and the error would occur.

Fixing it reduced the sync time from ~35s to ~2s; similar to the time it takes to sync the transactions.

**Review by Klakurka — APPROVED** [review]
I still get the blank page for a bit (with nav showing) but now it's only like 1-2 seconds so I'm good with that.

**Review by ScottMcDermid — APPROVED** [review]
Looks good!
